src/async-wrap.cc

```diff
@@ -177,6 +179,135 @@ static void PushBackDestroyId(Environment* env, double id) {
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(wrap->env());
+      FatalException(wrap->env()->isolate(), try_catch);
+      return false;
+    }
+  }
+
+  if (wrap->env()->using_domains()) {
+    Local<Value> domain_v = wrap->object()->Get(wrap->env()->domain_string());
+    if (domain_v->IsObject()) {
+      Local<Object> domain = domain_v.As<Object>();
+      if (domain->Get(wrap->env()->disposed_string())->IsTrue())
+        return false;
+      Local<Value> exit_v = domain->Get(wrap->env()->exit_string());
+      if (exit_v->IsFunction()) {
+        if (exit_v.As<Function>()->Call(domain, 0, nullptr).IsEmpty()) {
+          FatalError("node::AsyncWrap::MakeCallback",
+                    "domain exit callback threw, please report this");
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+class PromiseWrap : public AsyncWrap {
+ public:
+  PromiseWrap(Environment* env, Local<Object> object)
+    : AsyncWrap(env, object, PROVIDER_PROMISE) {}
+  size_t self_size() const override { return sizeof(*this); }
+};
+
+
+static void GetPromiseDomain(Local<v8::Name> name,
+                             const v8::PropertyCallbackInfo<v8::Value>& info) {
+  Local<Context> context = info.GetIsolate()->GetCurrentContext();
+  Environment* env = Environment::GetCurrent(context);
+  info.GetReturnValue().Set(Object::Cast(*info.Data())->Get(context,
+                            env->domain_string()).ToLocalChecked());
+}
+
+
+static void PromiseHook(PromiseHookType type, Local<Promise> promise,
+                        Local<Value> parent) {
+  Local<Context> context = promise->CreationContext();
+  Environment* env = Environment::GetCurrent(context);
+  const char async_id_key[] = "__async_wrap";
+  const char tag_id_key[] = "__async_wrap_tag";
+  if (type == PromiseHookType::kInit) {
+    // Unfortunately, promises don't have internal fields. Need a surrogate that
+    // async wrap can wrap.
+    Local<v8::ObjectTemplate> tem = v8::ObjectTemplate::New(env->isolate());
+    tem->SetInternalFieldCount(1);
+    Local<Object> obj = tem->NewInstance(context).ToLocalChecked();
+    PromiseWrap* wrap = new PromiseWrap(env, obj);
+    promise->DefineOwnProperty(context,
+              FIXED_ONE_BYTE_STRING(env->isolate(), async_id_key),
+              v8::External::New(env->isolate(), wrap),
+              v8::PropertyAttribute::DontEnum).FromJust();
+    promise->DefineOwnProperty(context,
+              FIXED_ONE_BYTE_STRING(env->isolate(), tag_id_key),
+              obj, v8::PropertyAttribute::DontEnum).FromJust();
+    if (env->in_domain()) {
+      obj->Set(context, env->domain_string(),
+        env->domain_array()->Get(context, 0).ToLocalChecked()).FromJust();
+      promise->SetAccessor(context, env->domain_string(),
+                           GetPromiseDomain, NULL, obj).FromJust();
+    }
+  } else if (type == PromiseHookType::kResolve) {
+    // TODO(matthewloring): need to expose this through the async hooks api.
```

Comment:
> This approach does not allow async_hook users to "react" to a promise resolution (state cannot be observed/updated at resolution time). Would it be too much to add another async_hook type for the resolution of async work (promise or otherwise)? With the addition of the C++ embedder api I can imagine that other types of async work could notify when async operations completes even if the JS callback isn't invoked immediately.

While it is not a technical issue to add another type (I assume you mean event type and not resource type), I think it would be better to do in a future PR where we have a better chance to evaluate the needs for such event type for other cases than promises.

So far the only concern mentioned in the EPS has been regarding context across async boundaries, which "attaching the async id to the promise resource" does solve, although not in a particularly nice way.